File: 10_data_management/agentr/R/guardrails.R
# R/guardrails.R
# Limits and safe paths for the course autonomous agent (Plumber edition)
# Tim Fraser
#
# Topic: AI for Data Management — parity with agentpy/app/guardrails.py

`%||%` = function(a, b) {
  if (is.null(a)) {
    b
  } else {
    a
  }
}

# 0. CONSTANTS ###############################################################

MAX_AUTONOMOUS_TURNS = 10L
MAX_WEB_SEARCHES_PER_REQUEST = 3L
MAX_SKILL_READS_PER_REQUEST = 8L

SKILLS_DIR_NAME = "skills"

# 1. PATHS ###################################################################

#' Activity root: AGENT.md, skills/, logs/ (agentr/ folder when cwd is set correctly).
agent_root = function() {
  o = Sys.getenv("AGENTR_ROOT", unset = "")
  if (nzchar(o)) {
    return(normalizePath(o, winslash = "/", mustWork = FALSE))
  }
  normalizePath(getwd(), winslash = "/", mustWork = FALSE)
}

skills_dir = function() {
  file.path(agent_root(), SKILLS_DIR_NAME)
}

# 2. SKILL FILES #############################################################

skill_basename_ok = function(basename) {
  grepl("^[a-zA-Z0-9_-]+\\.md$", basename)
}

read_skill_file = function(basename) {
  if (is.null(basename) || !nzchar(as.character(basename))) {
    stop("Skill name must be a non-empty string.")
  }
  basename = as.character(basename)
  if (!skill_basename_ok(basename)) {
    stop("Invalid skill filename (use basename like evidence_brief.md).")
  }
  full = normalizePath(file.path(skills_dir(), basename), winslash = "/", mustWork = FALSE)
  root_resolved = normalizePath(skills_dir(), winslash = "/", mustWork = FALSE)
  if (!startsWith(full, root_resolved)) {
    stop("Skill path must stay inside skills/.")
  }
  if (!file.exists(full)) {
    stop(paste0("Skill not found: ", basename))
  }
  paste(readLines(full, encoding = "UTF-8", warn = FALSE), collapse = "\n")
}

clamp_turns = function(requested) {
  if (is.null(requested)) {
    return(as.integer(MAX_AUTONOMOUS_TURNS))
  }
  n = suppressWarnings(as.integer(requested))
  if (is.na(n)) {
    return(as.integer(MAX_AUTONOMOUS_TURNS))
  }
  max(1L, min(n, as.integer(MAX_AUTONOMOUS_TURNS)))
}

min_completion_turns = function() {
  raw = trimws(Sys.getenv("AGENT_MIN_COMPLETION_TURNS", unset = ""))
  if (nzchar(raw) && grepl("^[0-9]+$", raw)) {
    n = as.integer(raw)
    return(max(1L, min(n, as.integer(MAX_AUTONOMOUS_TURNS))))
  }
  if (nzchar(trimws(Sys.getenv("SERPER_API_KEY", unset = "")))) {
    return(min(2L, as.integer(MAX_AUTONOMOUS_TURNS)))
  }
  1L
}

task_size_ok = function(task, max_chars = NULL) {
  if (is.null(max_chars)) {
    mc = Sys.getenv("AGENT_MAX_TASK_CHARS", unset = "8000")
    max_chars = suppressWarnings(as.integer(mc))
    if (is.na(max_chars)) {
      max_chars = 8000L
    }
  }
  is.character(task) && length(task) == 1L && nzchar(task) && nchar(task) <= max_chars
}

File: 10_data_management/agentpy/app/guardrails.py
# guardrails.py
# Limits and safe paths for the course autonomous agent (used by loop.py)
# Tim Fraser

import os
import re
from pathlib import Path

# Topic: AI for Data Management — keep guardrails obvious and readable for systems engineers.

MAX_AUTONOMOUS_TURNS = 10
MAX_WEB_SEARCHES_PER_REQUEST = 3
MAX_SKILL_READS_PER_REQUEST = 8

# Activity root: parent of this package (AGENT.md, skills/, logs/ live here).
_SKILLS_DIR_NAME = "skills"

# Skill files: basename only, alphanumeric, underscore, hyphen, .md
_SKILL_BASENAME_PATTERN = re.compile(r"^[a-zA-Z0-9_-]+\.md$")


def agent_root() -> Path:
    """Directory containing AGENT.md and skills/ (parent of the app/ package)."""
    return Path(__file__).resolve().parent.parent


def skills_dir() -> Path:
    """Directory containing markdown skill files loaded via read_skill."""
    return agent_root() / _SKILLS_DIR_NAME


def read_skill_file(basename: str) -> str:
    """
    Read skills/{basename} if basename matches ^[a-zA-Z0-9_-]+\\.md$.
    Raises ValueError if invalid or file missing.
    """
    if not basename or not isinstance(basename, str):
        raise ValueError("Skill name must be a non-empty string.")
    if not _SKILL_BASENAME_PATTERN.fullmatch(basename):
        raise ValueError("Invalid skill filename (use basename like evidence_brief.md).")
    full = (skills_dir() / basename).resolve()
    try:
        full.relative_to(skills_dir().resolve())
    except ValueError as exc:
        raise ValueError("Skill path must stay inside skills/.") from exc
    if not full.is_file():
        raise ValueError(f"Skill not found: {basename}")
    return full.read_text(encoding="utf-8")


def clamp_turns(requested: int | None) -> int:
    """Clamp requested max turns to [1, MAX_AUTONOMOUS_TURNS]."""
    if requested is None:
        return MAX_AUTONOMOUS_TURNS
    try:
        n = int(requested)
    except (TypeError, ValueError):
        return MAX_AUTONOMOUS_TURNS
    return max(1, min(n, MAX_AUTONOMOUS_TURNS))


def min_completion_turns() -> int:
    """
    Minimum LLM /api/chat rounds (each counts toward the turn budget) before END_BRIEF is accepted.

    - Set **AGENT_MIN_COMPLETION_TURNS** to a positive integer (capped at **MAX_AUTONOMOUS_TURNS**).
    - If unset: **2** when **SERPER_API_KEY** is set (nudge verification against search), else **1**.
    """
    raw = (os.getenv("AGENT_MIN_COMPLETION_TURNS") or "").strip()
    if raw.isdigit():
        n = int(raw)
        return max(1, min(n, MAX_AUTONOMOUS_TURNS))
    if (os.getenv("SERPER_API_KEY") or "").strip():
        return min(2, MAX_AUTONOMOUS_TURNS)
    return 1


def task_size_ok(task: str, max_chars: int | None = None) -> bool:
    """Optional coarse limit on prompt injection / huge payloads."""
    limit = max_chars if max_chars is not None else int(os.getenv("AGENT_MAX_TASK_CHARS", "8000"))
    return isinstance(task, str) and 0 < len(task) <= limit
